clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1202,6 +1246,84 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
   }
 };
 
+// A type to check the validity of passing union with
+// pointer/accessor/sampler/stream member as a kernel argument types.
+class SyclKernelUnionChecker : public SyclKernelFieldHandler {
+  int UnionCount = 0;
+  bool IsInvalid = false;
+  DiagnosticsEngine &Diag;
+
+public:
+  SyclKernelUnionChecker(Sema &S)
+      : SyclKernelFieldHandler(S), Diag(S.getASTContext().getDiagnostics()) {}
+  bool isValid() { return !IsInvalid; }
+  static constexpr const bool VisitUnionBody = true;
+
+  bool enterUnion(const CXXRecordDecl *RD, FieldDecl *FD) {
+    ++UnionCount;
+    return true;
+  }
+
+  bool leaveUnion(const CXXRecordDecl *RD, FieldDecl *FD) {
+    --UnionCount;
+    return true;
+  }
+
+  bool handlePointerType(FieldDecl *FD, QualType FieldTy) final {
+    if (UnionCount) {
+      IsInvalid = true;
+      Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
```

Comment:
You need a new diagnostic that says about union members. This one is incorrect here.

Reply:
Done.